sync/class.jetpack-sync-actions.php

```diff
@@ -53,11 +64,52 @@ static function init() {
 		if ( apply_filters( 'jetpack_sync_listener_should_load', true ) ) {
 			self::initialize_listener();
 		}
-		
+
 		add_action( 'init', array( __CLASS__, 'add_sender_shutdown' ), 90 );
 
 	}
 
+	static function maybe_start_initial_sync() {
+		if ( ! self::can_do_initial_sync() ) {
+			return;
+		}
+		// Previous
+		$previous_version_and_time = Jetpack_Options::get_option( 'old_version', 0 );
+		$previous_version          = explode( ':', $previous_version_and_time );
+		self::do_initial_sync( JETPACK__VERSION, $previous_version[0], true );
+	}
+
+	static function can_do_initial_sync( $current_blog_id = null, $current_time = null ) {
+		if ( empty( $current_blog_id ) ) {
+			$current_blog_id = get_current_blog_id();
+		}
+		if ( empty( $current_time ) ) {
+			$current_time = time();
+		}
+
+		$version_with_time = explode( ':', Jetpack_Options::get_option( 'version', 0 ) );
+		if ( ! isset( $version_with_time[1] ) ) {
+			// This is not very likely to happen.
+			// lets set it to 0 so that the update happends right away
+			$version_with_time[1] = 0;
+		}
+		$version_updated = $version_with_time[1];
+
+		/**
+		 * Allows the dev to change the number of blogs that the nework is allowed update per second.
+		 * By default this value is set to 10 blogs per second.
+		 * The blogs blog_id determins if a site can update.
+		 *
+		 * @since 4.5.0
+		 *
+		 * @param int the number of blogs per second that should be allowed to update.
+		 */
+		$blogs_per_seconds = (int) apply_filters( 'jetpack_network_ramp_up_blogs_per_second', self::NETWORK_UPDATE_RAMP_UP_BLOGS_PER_SECOND );
+		$time_difference   = ( $current_time - $version_updated );
+
+		return ( $current_blog_id <= ( $time_difference * $blogs_per_seconds ) );
```

Comment:
I don't know quite how this is supposed to only sync 10 blogs per second. It seems to me that it's highly dependent on the distribution of blog_ids. For example, on WPCOM we have blog IDs up to 100,000,000 which would mean that with this algorithm the last blog would sync in 115 days after upgrade, but in between there would be lots of huge gaps, perhaps days-long, in which no blogs are synced at all.

I'm not sure this is the best way to implement this. It might be better to track how many blogs have been upgraded and save that as a variable, or to have the upgrade happen in a single process which iterates through 10 non-upgraded blogs in a given wp_cron cycle.